Make pointer and selected sign customizable

relastle opened the pull request:
Thank you for developing and maintaining this state-of-the-art CLI.

I opened this PR because as well as prompt, pointer sign and selected signs (in multi-select mode) should also be customizable with command line arguments.

![fzf_demo](https://user-images.githubusercontent.com/6816040/73324272-c91a7e00-428d-11ea-9b20-d5c8ab327b92.gif)

Such feature, for example, would be useful in a case, where UI-provider really want users to know "WHAT" will happen ( `multi-delete` :fire: or `multi-star` ⭐️ ) with emoji.

relastle commented:
Sorry, this PR is currently behind the upstream HEAD, i will fix it now.

junegunn commented:
Thanks for your interest in the project. This was suggested a few times in the past, but I rejected the idea because I felt that adding it would eventually lead to feature creep. You can read my previous comments on that matter:

- https://github.com/junegunn/fzf/issues/692#issuecomment-251436739
- https://github.com/junegunn/fzf/issues/1769#issuecomment-557946235

Fast forward a few years, I'm not sure if I still feel the same. But you made a good point that using different marker signs can provide additional context to the users, so that this is not purely cosmetic.

There are a few things I'd do differently:

- We already use the terms "pointer" and "marker" (see `man fzf`), so I would go for more concise option names: `--pointer` and `--marker`
- I'd like to limit the width of the signs to 2 (Emojis or CJK wide characters are allowed), and make sure that they don't contain `\n` or `\t` so we have less to worry about screen overflow issues: (`fzf --pointer "$(seq 100)"`)
- Make it compatible with `jump` feature.
- Test cases would be nice.

relastle commented:
@junegunn 

Thank you for the insightful review comments.

>We already use the terms "pointer" and "marker" (see man fzf), so I would go for more concise option names: --pointer and --marker

I'm sorry for not following the existing naming concistency.
It was fixed in (https://github.com/junegunn/fzf/pull/1844/commits/2aff65ccafcd306946e439c1793f9d0301dd1e6d)

> I'd like to limit the width of the signs to 2 (Emojis or CJK wide characters are allowed), and make sure that they don't contain \n or \t so we have less to worry about screen overflow issues: (say no to fzf --pointer "$(seq 100)")

That's really reasonable. I added option parameter validation and some table-driven test-cases (https://github.com/junegunn/fzf/pull/1844/commits/c54dbc30d13a7636abc422efecdf74d6d8eda521) 

> Make it compatible with jump feature.

I forgot to test new functions working with that feature.
It was fixed in (https://github.com/junegunn/fzf/pull/1844/commits/df6b1441f29ffdf32c79c75100a6d2a169c5d81f)

-- -

Finally, I personally don't think that `pointer` and `marker` display length should be absolutely less than or equal `2`. (i.g. Someone may want to use `>>>` as pointer 🤔 )
So I implemented the logics so that only few changes should be done when the limit of display length would be longer (than 2) in the future.

relastle commented:
![fzf_jump_demo](https://user-images.githubusercontent.com/6816040/74601349-93282700-50e0-11ea-88c5-f6169d87951d.gif)

Here is the demonstration of multi-display-width marker and pointer working with the jump feature.

junegunn commented on man/man1/fzf.1:
```diff
@@ -231,6 +231,12 @@ A synonym for \fB--info=hidden\fB
+Pointer sign of cursor line (default: '>')
+.TP
+.BI "--marker=" "STR"
+Signs of selected ling in multi-mode  (default: '>')
```
Let's just use the same expressions in `--color` section of the man page.

```
pointer    Pointer to the current line
marker     Multi-select marker
```

junegunn commented on shell/completion.bash:
```diff
@@ -72,6 +72,8 @@ _fzf_opts_completion() {
     --inline-info
     --prompt
+    --pointer
+    --selected
```
`--marker`

junegunn commented on src/options.go:
```diff
@@ -1333,6 +1354,33 @@ func parseOptions(opts *Options, allArgs []string) {
+}
+
+func validateSign(sign string, signOptName string) error {
+	if strings.Contains(sign, "\n") || strings.Contains(sign, "\t") {
```
It would be more precise to use https://golang.org/pkg/unicode/#IsGraphic
Error message can be `invalid character in %v`

junegunn commented on src/options.go:
```diff
@@ -59,6 +60,8 @@ const usage = `usage: fzf [options]
     --info=STYLE          Finder info style [default|inline|hidden]
     --prompt=STR          Input prompt (default: '> ')
+    --pointer=STR         Cursor line pointer sign (default: '>')
+    --marker=STR          Selected sign in multi-select mode (default: '>')
```
Also, the same expressions:

```
Pointer to the current line (default: '>')
Multi-select marker (default: '>')
```

relastle commented:
@junegunn 

Thank you for your reviews!

I amended the description of the options and added three integration test-cases.

junegunn commented on src/options.go:
```diff
@@ -1333,6 +1355,32 @@ func parseOptions(opts *Options, allArgs []string) {
+	}
+	if widthSum > 2 {
+		return fmt.Errorf("%v display width should be up to 2", signOptName)
+	}
```
We can put `if` statement inside `for` loop for early termination.

relastle replied:
Thank you.

fixed in (ddc7435)

relastle commented:
> We should disallow empty pointer or marker.

Exactly. Validation for empty string and its test-case was added in (5b8bca2)

junegunn commented:
Merged, thanks a lot!

relastle commented:
Thank you, too!
